File: 0835-image-overlap/0835-image-overlap.go
func largestOverlap(img1 [][]int, img2 [][]int) int {
	n := len(img1)

	// Convert images to bitmaps
	var image1 [30]uint32
	var image2 [30]uint32
	for i := range img1 {
		for j := range img1 {
			image1[i] = (image1[i] << 1) | uint32(img1[i][j])
			image2[i] = (image2[i] << 1) | uint32(img2[i][j])
		}
	}

	// For each possible shift in x and y axis, try it and count common bits
	var res int
	for dy := -n + 1; dy < n; dy++ {
		for dx := 0; dx < n; dx++ {
			var left int
			var right int
			for i := 0; i < n; i++ {
				j := i + dy
				if j < 0 || j >= n {
					continue
				}
				// Note: we do not need to worry about going out of bounds
				// because the other image could not share those bits.
				left += bits.OnesCount32(image2[i] & (image1[j] << dx))
				right += bits.OnesCount32(image2[i] & (image1[j] >> dx))
			}
			res = max(res, left)
			res = max(res, right)
		}
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 0835-image-overlap/0835-image-overlap.ts
function largestOverlap(img1: number[][], img2: number[][]): number {
    const n = img1.length;
    let best = 0;

    for (let dx = -n + 1; dx <= n - 1; dx++) {
        for (let dy = -n + 1; dy <= n - 1; dy++) {
            let score = 0;
            for (let i = 0; i < n; i++) {
                for (let j = 0; j < n; j++) {
                    let c = 0;
                    if (dx + i >= 0 && dx + i < n && dy + j >= 0 && dy + j < n) {
                        c = img1[dx + i][dy + j];
                    }
                    score += img2[i][j] & c;
                }
            }
            best = Math.max(best, score);
        }
    }

    return best;
}